### PR TITLE
Add Hilt dependency injection framework

### LINKED ISSUE #103: Add Hilt dependency injection to our application
- Will allow us to do mock testing in the future

### PR DESCRIPTION
**Pull request by armiantos**
# Summary
- Use DI to enable the possibility of Mock testing in the future
- Uses [Android Hilt](https://developer.android.com/training/dependency-injection/hilt-android)
- Also fixes the missing setter in `Request.java` (Firestore would return no book reference otherwise, will also be done by @Taranjot-Singh  in his unlibrary query PR)

## How to use

Fragment classes that require Hilt dependencies should be marked with `@AndroidEntryPoint` to let Hilt know 'Hey give me these dependencies'. The activity containing this fragment must also be marked with `@AndroidEntryPoint`. 

When you're making a Fragment that uses a ViewModel, use `new ViewModelProvider(requireActivity()).get(ViewModelName.class)` to get the view model during `onCreate()`. Refer to [this](https://developer.android.com/training/dependency-injection/hilt-jetpack) for more information.

The dependency objects are marked with `@Inject` in their constructors to let Hilt know, here's how to provide this type of instance. ViewModels have a special `@ViewModelInject` annotation to put in their constructors. [ref](https://developer.android.com/training/dependency-injection/hilt-jetpack)

When your ViewModel object requires a Repository. Annotate the Repository constructor with `@Inject` and add a repository instance in your ViewModel constructor. Hilt will get the repository for you. 

If your repositories have dependencies that have constructor or instances outside the code you write (we can't mark these object's constructors with `@Inject` ) create a Module provider. Find more information about this [here](https://developer.android.com/training/dependency-injection/hilt-android#inject-provides)

Dependencies are listed in the constructor of the class needing it.

# TODO
- Create mock repositories for tests in the future

Closes #103

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -19,6 +21,7 @@
     </queries>
 
     <application
+        android:name=".MainApplication"
```
Comment (armiantos) [inline app/src/main/AndroidManifest.xml:24]: In order to use Hilt, we must provide an Application class marked with `@HiltAndroidApp`. [ref](https://developer.android.com/training/dependency-injection/hilt-android#application-class)

File: app/src/main/java/com/example/unlibrary/MainActivity.java [hunk 2]
```diff
@@ -28,9 +28,12 @@
 import java.util.Date;
 import java.util.Locale;
 
+import dagger.hilt.android.AndroidEntryPoint;
+
 /**
  * Host Activity for the app
  */
+@AndroidEntryPoint
```
Comment (armiantos) [inline app/src/main/java/com/example/unlibrary/MainActivity.java:36]: Activity that contains Fragments marked with `@AndroidEntryPoint` must be marked with `@AndroidEntryPoint` as well. 
> If you annotate an Android class with @AndroidEntryPoint, then you also must annotate Android classes that depend on it. For example, if you annotate a fragment, then you must also annotate any activities where you use that fragment.

File: app/src/main/java/com/example/unlibrary/exchange/ExchangeBookDetailsFragment.java [hunk 3]
```diff
@@ -23,9 +23,12 @@
 
 import org.jetbrains.annotations.NotNull;
 
+import dagger.hilt.android.AndroidEntryPoint;
+
 /**
  * Fragment to display the details and requests upon a user owned book.
  */
+@AndroidEntryPoint
```
Comment (armiantos) [inline app/src/main/java/com/example/unlibrary/exchange/ExchangeBookDetailsFragment.java:31]: > Hilt can provide dependencies to other Android classes that have the @AndroidEntryPoint annotation:

File: app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java [hunk 4]
```diff
@@ -49,8 +49,9 @@
     /**
      * Constructor for the Exchange Repository.
      */
-    public ExchangeRepository() {
-        mDb = FirebaseFirestore.getInstance();
+    @Inject
```
Comment (armiantos) [inline app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java:52]: > Use the @Inject annotation on the constructor of a class to tell Hilt how to provide instances of that class

File: app/src/main/java/com/example/unlibrary/exchange/ExchangeViewModel.java [hunk 5]
```diff
@@ -42,8 +43,9 @@
     /**
      * Constructor for the ExchangeViewModel. Instantiates listener to Firestore.
      */
-    public ExchangeViewModel() {
-        mExchangeRepository = new ExchangeRepository();
+    @ViewModelInject
```
Comment (armiantos) [inline app/src/main/java/com/example/unlibrary/exchange/ExchangeViewModel.java:46]: > Provide a ViewModel using the @ViewModelInject annotation in the ViewModel object's constructor. You must also annotate the SavedStateHandle dependency with @Assisted

**Review by armiantos — COMMENTED** [review]
Here's a general guideline how to use the Hilt annotations

File: app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java [hunk 6]
```diff
@@ -49,8 +49,9 @@
     /**
      * Constructor for the Exchange Repository.
      */
-    public ExchangeRepository() {
-        mDb = FirebaseFirestore.getInstance();
+    @Inject
+    public ExchangeRepository(FirebaseFirestore db) {
```
Comment (armiantos) [inline app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java:53]: Dependency objects should be listed in the constructor of the class that needs it.

File: app/src/main/java/com/example/unlibrary/unlibrary/UnlibraryViewModel.java [hunk 7]
```diff
@@ -29,6 +30,7 @@
      * Constructor for the UnLibrary ViewModel. Binds the list of books to return from the
      * repository.
      */
+    @ViewModelInject
     public UnlibraryViewModel() {
         mUnlibraryRepository = new UnlibraryRepository();
```
Comment (cyrusdiego) [inline app/src/main/java/com/example/unlibrary/unlibrary/UnlibraryViewModel.java:35]: looks like you missed a DI for the unlibrary repository :P

File: app/src/main/java/com/example/unlibrary/unlibrary/UnlibraryRepository.java [hunk 8]
```diff
@@ -35,6 +37,7 @@
      * <p>
      * TODO: Add querying logic to return only books that have been requested or borrowed by the user
      */
+    @Inject
     public UnlibraryRepository() {
```
Comment (cyrusdiego) [inline app/src/main/java/com/example/unlibrary/unlibrary/UnlibraryRepository.java:41]: inject database instance

File: app/src/main/java/com/example/unlibrary/unlibrary/UnlibraryViewModel.java (hunk 7 above)
Comment (armiantos) [reply app/src/main/java/com/example/unlibrary/unlibrary/UnlibraryViewModel.java:35]: Ah wait hahaha good catch

Comment (armiantos) [reply app/src/main/java/com/example/unlibrary/unlibrary/UnlibraryViewModel.java]: Fixed!

File: app/src/main/java/com/example/unlibrary/unlibrary/UnlibraryRepository.java (hunk 8 above)
Comment (armiantos) [reply app/src/main/java/com/example/unlibrary/unlibrary/UnlibraryRepository.java]: Thanks and fixed!

**Review by cyrusdiego — APPROVED** [review]
LGTM! :D code got a lot cleaner haha

**Comment by armiantos** [comment]
Sweet! I'll merge this once Daniels' and Taran's PR go through.

**Comment by armiantos** [comment]
Added `@AndroidEntryPoint` to @drojasca's `UnlibraryBookDetailsFragment`

**Review by cyrusdiego — APPROVED** [review]
LGTM! :D

**Pull request merged**